File: jvm/src/jmh/java/com/github/jcornaz/collekt/ListImplementation.java
package com.github.jcornaz.collekt;

import com.github.jcornaz.collekt.impl.KotlinList;

public enum ListImplementation {
    KOTLIN_LIST(KotlinList.Factory);

    public final PersistentListFactory factory;

    ListImplementation(PersistentListFactory factory) {
        this.factory = factory;
    }
}

File: jvm/src/jmh/kotlin/com/github/jcornaz/collekt/ListImplementation.kt
package com.github.jcornaz.collekt

import com.github.jcornaz.collekt.impl.KotlinList
import com.github.jcornaz.collekt.impl.PaguroRRBTree

enum class ListImplementation(protected val factory: PersistentListFactory) : PersistentListFactory by factory {
    KOTLIN_LIST(KotlinList) {
        override fun <E> from(iterable: Iterable<E>): PersistentList<E> =
                factory.from(iterable) // faster creation as KotlinList cannot be unbalanced
    },

    PAGURO_RRB_TREE(PaguroRRBTree)

    ;

    // Use fold instead of delegating to factory.from() in order to create unbalanced data structure when relevant
    override fun <E> from(iterable: Iterable<E>): PersistentList<E> =
            iterable.fold(factory.empty()) { result, element -> result + element }
}
